tests/cbmc/proofs/s2n_dh_params_free/s2n_dh_params_free_harness.c

```diff
@@ -30,8 +30,17 @@ void s2n_dh_params_free_harness()
     nondet_s2n_mem_init();
 
     /* Operation under verification. */
-    if (s2n_dh_params_free(dh_params) == S2N_SUCCESS) {
-        /* Postconditions. */
+    int result = s2n_dh_params_free(dh_params);
+
+    /* Postconditions. */
+    if (result == S2N_SUCCESS) {
         assert(dh_params->dh == NULL);
+    } else {
+        assert(dh_params == NULL && s2n_errno == S2N_ERR_NULL);
     }
```

Comment:
Why prove that the only reason we fail is S2N_ERR_NULL?

Reply:
It just makes the proof complete.

I can remove the assertion if you think it's not needed, but for deallocation failures, it might be good to document the expected failures. Let me know.

Reply:
Are you doing this because there's currently only one failure case? If there were two or more possible ways for this method to fail, would you be checking that at least one was true?

Reply:
Also, is s2n_dh_params_free failing on NULL actually the right behavior here, or should it be a no-op?

Reply:
> If there were two or more possible ways for this method to fail, would you be checking that at least one was true?

Yes, ideally we would have the following structure:

```c
if (s2n_errno = ERR_1) {
  assert(... condition under which ERR_1 is expected ...);
} else (s2n_errno = ERR_2) {
  assert(... condition under which ERR_2 is expected ...);
}
```

> Also, is `s2n_dh_params_free` failing on `NULL` actually the right behavior here, or should it be a no-op?

This is a good question. IMO, it should be a no-op. If we agree, then I could remove this assertion and we can track this proposal in a separate GitHub issue.
At the end of the day, it's a design decision for the implementation, and what I am trying to capture here is a specification for the function.

Reply:
Removed this assertion and opened #2802.

Reply:
I worry that model could easily get out of hand for any mildly complex function :\ If you look at our code, any RESULT/POSIX_GUARD represents a possible failure case. Almost every line could fail.

But yes, I don't see any issue with s2n_dh_params_free(NULL) being a no-op.

Reply:
> But yes, I don't see any issue with s2n_dh_params_free(NULL) being a no-op.

Thanks for checking. I wouldn't make the implementation change to `s2n_dh_params_free` as part of this PR though, since now that I have removed the assertion, the no-memory-leak proof would go through regardless. #2802 can be closed in a separate PR.

> Wouldn't it be a more useful proof to assert things that should always be true when the method fails, regardless of how it fails?

It would be less overhead, but the proof would be an overapproximation -- just like this proof currently is: it doesn't care if `s2n_dh_params_free` fails on `NULL` or not.
I was trying to write a tight specification for the function. You may be right -- once we start looking at more complex functions, the case splits might be really hard to write. I don't have a good intuition for that yet.